core/src/main/java/hudson/model/Queue.java

```diff
@@ -1631,17 +1632,23 @@ public void maintain() {
                 } else {
 
                     List<JobOffer> candidates = new ArrayList<>(parked.size());
-                    List<CauseOfBlockage> reasons = new ArrayList<>(parked.size());
+                    Map<Node, CauseOfBlockage> reasonMap = new HashMap<>();
                     for (JobOffer j : parked.values()) {
-                        CauseOfBlockage reason = j.getCauseOfBlockage(p);
+                        Node offerNode = j.getNode();
+                        CauseOfBlockage reason;
+                        if (reasonMap.containsKey(offerNode)) {
+                            reason = reasonMap.get(offerNode);
+                        } else {
+                            reason = j.getCauseOfBlockage(p);
+                            reasonMap.put(offerNode, reason);
+                        }
```

Comment:
Putting null to reasonMap is expected IIUC

```suggestion
                        CauseOfBlockage reason = reasonMap.get(offerNode);
                        if (reason == null) {
                            reason = j.getCauseOfBlockage(p);
                            reasonMap.put(offerNode, reason);
                        }
```

Reply:
Its ok to have nulls as values as the reason can be null so the suggested version will call getCauseOfBlockage more than once

Reply:
OK